File: 3/solution.py
# Path: 3/solution.py

chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"


def part1(input: str) -> int:
    input_list = input.split("\n")

    total = 0

    for i in input_list:
        # split i in half
        first_half = i[0 : int(len(i) / 2)]
        second_half = i[int(len(i) / 2) :]
        for c in chars:
            if first_half.count(c) > 0 and second_half.count(c) > 0:
                total += chars.index(c) + 1

    return total


def part2(input: str) -> int:
    input_list = input.split("\n")

    total = 0

    for i in range(0, len(input_list), 3):
        first = input_list[i]
        second = input_list[i + 1]
        third = input_list[i + 2]
        for c in chars:
            if first.count(c) > 0 and second.count(c) > 0 and third.count(c) > 0:
                total += chars.index(c) + 1

    return total


if __name__ == "__main__":
    input = open("./3/input.txt", "r").read()
    print("Part 1: " + str(part1(input)))
    print("Part 2: " + str(part2(input)))

File: 3/solution.rs
// Path: 3/solution.rs

use std::{fs::File, io::Read};

const CHARS: &str = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";

fn part1(input: &str) -> usize {
    let input: Vec<&str> = input.split("\n").collect();

    let mut sum = 0;

    for i in input {
        let first_half = i[0..(i.len() / 2)].to_string();
        let second_half = i[(i.len() / 2)..].to_string();

        CHARS.chars().for_each(|c| {
            if first_half.matches(c).count() > 0 && second_half.matches(c).count() > 0 {
                sum += CHARS.find(c).unwrap() + 1;
            }
        });
    }

    sum
}

fn part2(input: &str) -> usize {
    let input: Vec<&str> = input.split("\n").collect();

    let mut sum = 0;

    for i in (0..input.len()).step_by(3) {
        let first = input[i];
        let second = input[i + 1];
        let third = input[i + 2];

        CHARS.chars().for_each(|c| {
            if first.matches(c).count() > 0
                && second.matches(c).count() > 0
                && third.matches(c).count() > 0
            {
                sum += CHARS.find(c).unwrap() + 1;
            }
        });
    }

    sum
}

fn read_input() -> String {
    let mut input = String::new();
    let mut file = File::open("./3/input.txt").unwrap();
    file.read_to_string(&mut input).unwrap();
    input
}

fn main() {
    let input = read_input();
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
